### ISSUE TITLE
Library panics when deserializing certain types

### ISSUE THREAD

**Issue by ryan-summers** [rust-embedded-community/serde-json-core#48]
When deserializing certain types, the library may hit `unreachable!()` statements, which results in a panic. This appears to be because serde-json-core does not support deserializing complex enums (e.g. `pub struct Test(u32)`.

That being said, the library should never internally panic. Instead, it should propagate an error out to the user.

**Comment by dnadlinger** [comment]
(The particular issue of `newtype` has been fixed in https://github.com/rust-embedded-community/serde-json-core/commit/cce045cf76eb4b60d08241cf0a2a2458b49189c1; not sure whether there are any others left.)

**Comment by jordens** [comment]
These are all known at compile time. Shouldn't already the `derive` fail?

**Comment by ryan-summers** [comment]
The derive is done in `serde`, which doesn't care about the support of the underlying serialize/deserialize engine. Hence, we are just provided a serde-compatible API at the serde-json-core level. We don't have any proc-macros in this repository, we just have functions that can deserialize/serialize objects that implement `serde::Serialize` and `serde::Deserialize`. In other backends for serde, these types are supported.

**Comment by ryan-summers** [comment]
Closing, as I don't think this has any open defects anymore with newtype resolved.

**Comment by jordens** [comment]
I think it's still pretty easy to hit, e.g. `serde_json_core::to_vec::<_, 10>(&'a')`

**Comment by gdemenibus** [comment]
Another type that hits unreachable is serialize_tuple_variant. Trying to serialize an enum with a tuple element will hit unreachable in ```src/mod.rs:407```

**Comment by gdemenibus** [comment]
You can also hit an unreachable section by serializing an enum with a tuple element.

**Comment by jordens** [comment]
I don't think this general behavior is a bug. It seems to be standard practice Form serde trait implementations to panic on unimplemented/unimplementable things. What would be the alternative?

**Comment by ryan-summers** [comment]
I think the core issue is that it's marked as `unreachable!()` when in reality it's an `unimplemented!()`. In either case, an internal panic is likely the only solution if we don't want to propogate an error up the stack.

**Comment by jordens** [comment]
The title should probably reflect that then.

**Comment by ryan-summers** [comment]
I believe this should be fixed now that #83 has made these functions return `Err` instead of `unreachable!()`. There's still some `unreachable!()` statements in `src/de/map.rs`.

I think the only case you'll now hit these is if you use some non-string as a map key, such as "`{0: "Test"}`

**Comment by sammhicks** [comment]
> I believe this should be mostly fixed now that #83 has made these functions return `Err` instead of `unreachable!()`. There's still some `unreachable!()` statements in `src/de/map.rs`.
> 
> I think the only case you'll now hit these is if you use some non-string as a map key, such as `{0: "Test"}`

These can easily be hit using map types (e.g. `HashMap` or `heapless::LinearMap`), so I think the logic should be as follows:

- Strings and chars are serialized normally.
- Numerical types, i.e. `i32` and `u32` are serialized normally, but surrounded by double quotes.
- All other types are rejected with an "key must be a string" error.

This has the benefit that it's more permissive, generally predictable, and mirrors what `serde_json` does.

I'm happy to submit a pull request if the design is approved.

**Comment by katyo** [comment]
I have same issue with the following:
https://github.com/rust-embedded-community/serde-json-core/blob/3b369a01d7d23195f1e0392ffb239e4602470567/src/ser/mod.rs#L594

**Comment by Kezii** [comment]
Bumping into this too

https://github.com/rust-embedded-community/serde-json-core/blob/a435f788c63fe41003170931ef09858c39b5a196/src/ser/mod.rs#L704
